### PR TITLE
Add test coverage for critical hits, recoil, and status effects etc

### PR DESCRIPTION
**Pull request by TurboRx**
Added a big batch of battle tests to cover various edge cases and mechanics. This ensures all the stat changes and battle interactions are working exactly as intended. 

A quick breakdown of what's covered:
* **Abilities:** `Verified boosts`, `exceptions`, and `Mold Breaker` interactions for `Adaptability`, `Bulletproof`, `Hustle`, damage reducers (like `Filter`), and pinch abilities `(Blaze, Torrent, etc.)`.
* **Items:** Ensured Choice items `(Band/Specs/Scarf)` correctly handle `stat boosts`, `move locking`, and `Dynamax` interactions.
* **Status Effects:** Added tests for `confusion self-hits`, `duration`, and prevention via `Own Tempo`.

### CODE REVIEW

**Comment by TurboRx** [comment]
Hey @andrebastosdias, lmk if you’ve already added the confusion tests in your PR. If you have, I’ll go ahead and remove them from mine to avoid any overlap.

File: test/sim/abilities/adaptability.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		], [
+			{ species: 'Blissey', ability: 'naturalcure', moves: ['sleeptalk'] },
+		]]);
+		battle.randomizer = dmg => dmg;
+		battle.makeChoices('move triattack', 'move sleeptalk');
+		const normalDamage = battle.p2.active[0].maxhp - battle.p2.active[0].hp;
+
+		// 2x / 1.5x ≈ 1.33x more damage
+		assert.bounded(adaptDamage / normalDamage, [1.28, 1.38]);
+	});
```
Comment (andrebastosdias) [inline test/sim/abilities/adaptability.js:24]: Why wouldn't you just put the damage in the calculator and check how much HP it lost? Forcing the randomizer and checking the ratio is weird.

File: test/sim/abilities/bulletproof.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	it(`should block ball and bomb moves`, () => {
+		battle = common.createBattle([[
+			{ species: 'Chesnaught', ability: 'bulletproof', moves: ['sleeptalk'] },
+		], [
+			{ species: 'Gengar', ability: 'levitate', moves: ['focusblast'] },
+		]]);
+		battle.makeChoices('move sleeptalk', 'move focusblast');
+		assert.fullHP(battle.p1.active[0]);
+		assert(battle.log.some(line => line.includes('|-immune|') && line.includes('Bulletproof')));
```
Comment (andrebastosdias) [inline test/sim/abilities/bulletproof.js]: You don't need to check for messages, you just need to check for damage. Messages can vary across time, they are not the preferred option.

File: test/sim/misc/confusion.js [hunk 3]
```diff
@@ -21,4 +21,42 @@ describe('Confusion', () => {
...
+			{ species: 'Sableye', ability: 'prankster', moves: ['confuseray', 'sleeptalk'] },
+		]]);
+		const snorlax = battle.p1.active[0];
+		battle.makeChoices('move sleeptalk', 'move confuseray');
+		assert(snorlax.volatiles['confusion']);
+		snorlax.volatiles['confusion'].time = 1;
+		battle.forceRandomChance = false;
+		battle.makeChoices('move sleeptalk', 'move sleeptalk');
+		assert(!snorlax.volatiles['confusion']);
+	});
```
Comment (andrebastosdias) [inline test/sim/misc/confusion.js:64]: Why are some of these tests needed?

File: test/sim/misc/critical.js [hunk 4]
```diff
@@ -21,4 +21,55 @@ describe('Critical hits', () => {
...
+			{ species: 'Octillery', ability: 'pressure', item: 'razorclaw', moves: ['watergun'] },
+		], [
+			{ species: 'Blissey', ability: 'naturalcure', moves: ['sleeptalk'] },
+		]]);
+		battle.randomizer = dmg => dmg;
+		battle.makeChoices('move watergun', 'move sleeptalk');
+		const critDamage = battle.p2.active[0].maxhp - battle.p2.active[0].hp;
+
+		assert(battle.log.some(line => line.startsWith('|-crit|')));
+		assert.bounded(critDamage / normalDamage, [1.45, 1.55]);
```
Comment (andrebastosdias) [inline test/sim/misc/critical.js]: Why do you keep checking damage like this?

File: test/sim/misc/sleep.js
Comment (andrebastosdias) [inline test/sim/misc/sleep.js:1]: There are already sleep tests in the statuses file.

**Review by andrebastosdias — COMMENTED** [review]
I'm sorry, but a lot of these seem forced for no reason.

**Comment by Slayer95** [comment]
I welcome all passing tests. And the more diverse test methods, the better.

File: test/sim/misc/confusion.js (hunk 3 above)
Comment (andrebastosdias) [reply test/sim/misc/confusion.js:64]: I don't want to be picky but:
- why are you setting `forceRandomChance` to false?
- if the test is supposed to check that the timer correctly descreases, why are you setting the timer to 1? It defeats the purpose of the test.

Choose a seed, run a loop `n` times checking that the Pokémon is still confused, and check that is not by the end of it.
